Implement minimal reflection probes.

pcwalton opened the pull request:
# Objective

This pull request implements *reflection probes*, which generalize environment maps to allow for multiple environment maps in the same scene, each of which has an axis-aligned bounding box. This is a standard feature of physically-based renderers and was inspired by [the corresponding feature in Blender's Eevee renderer].

## Solution

This is a minimal implementation of reflection probes that allows artists to define cuboid bounding regions associated with environment maps. For every view, on every frame, a system builds up a list of the nearest 4 reflection probes that are within the view's frustum and supplies that list to the shader. The PBR fragment shader searches through the list, finds the first containing reflection probe, and uses it for indirect lighting, falling back to the view's environment map if none is found. Both forward and deferred renderers are fully supported.

A reflection probe is an entity with a pair of components, *LightProbe* and *EnvironmentMapLight* (as well as the standard *SpatialBundle*, to position it in the world). The *LightProbe* component (along with the *Transform*) defines the bounding region, while the *EnvironmentMapLight* component specifies the associated diffuse and specular cubemaps.

A frequent question is "why two components instead of just one?" The advantages of this setup are:

1. It's readily extensible to other types of light probes, in particular *irradiance volumes* (also known as ambient cubes or voxel global illumination), which use the same approach of bounding cuboids. With a single component that applies to both reflection probes and irradiance volumes, we can share the logic that implements falloff and blending between multiple light probes between both of those features.

2. It reduces duplication between the existing *EnvironmentMapLight* and these new reflection probes. Systems can treat environment maps attached to cameras the same way they treat environment maps applied to reflection probes if they wish.

Internally, we gather up all environment maps in the scene and place them in a cubemap array. At present, this means that all environment maps must have the same size, mipmap count, and texture format. A warning is emitted if this restriction is violated. We could potentially relax this in the future as part of the automatic mipmap generation work, which could easily do texture format conversion as part of its preprocessing.

An easy way to generate reflection probe cubemaps is to bake them in Blender and use the `export-blender-gi` tool that's part of the [`bevy-baked-gi`] project. This tool takes a `.blend` file containing baked cubemaps as input and exports cubemap images, pre-filtered with an embedded fork of the [glTF IBL Sampler], alongside a corresponding `.scn.ron` file that the scene spawner can use to recreate the reflection probes.

Note that this is intentionally a minimal implementation, to aid reviewability. Known issues are:

* Reflection probes are basically unsupported on WebGL 2, because WebGL 2 has no cubemap arrays. (Strictly speaking, you can have precisely one reflection probe in the scene if you have no other cubemaps anywhere, but this isn't very useful.)

* Reflection probes have no falloff, so reflections will abruptly change when objects move from one bounding region to another.

* As mentioned before, all cubemaps in the world of a given type (diffuse or specular) must have the same size, format, and mipmap count.

Future work includes:

* Blending between multiple reflection probes.

* A falloff/fade-out region so that reflected objects disappear gradually instead of vanishing all at once.

* Irradiance volumes for voxel-based global illumination. This should reuse much of the reflection probe logic, as they're both GI techniques based on cuboid bounding regions.

* Support for WebGL 2, by breaking batches when reflection probes are used.

These issues notwithstanding, I think it's best to land this with roughly the current set of functionality, because this patch is useful as is and adding everything above would make the pull request significantly larger and harder to review.

---

## Changelog

### Added

* A new *LightProbe* component is available that specifies a bounding region that an *EnvironmentMapLight* applies to. The combination of a *LightProbe* and an *EnvironmentMapLight* offers *reflection probe* functionality similar to that available in other engines.

[the corresponding feature in Blender's Eevee renderer]: https://docs.blender.org/manual/en/latest/render/eevee/light_probes/reflection_cubemaps.html

[`bevy-baked-gi`]: https://github.com/pcwalton/bevy-baked-gi

[glTF IBL Sampler]: https://github.com/KhronosGroup/glTF-IBL-Sampler

pcwalton commented:
Both issues have been fixed. I want to improve the example before submitting this for review. This requires messing with the glTF IBL Sampler some more.

pcwalton commented:
This is ready for review.

pcwalton commented:
I pushed a new version of this patch that enables support for reflection probes in deferred rendering.

superdump commented on crates/bevy_pbr/src/environment_map/environment_map.wgsl:
```diff
@@ -16,14 +18,46 @@ fn environment_map_light(
-    let irradiance = textureSample(bindings::environment_map_diffuse, bindings::environment_map_sampler, vec3(N.xy, -N.z)).rgb;
-    let radiance = textureSampleLevel(bindings::environment_map_specular, bindings::environment_map_sampler, vec3(R.xy, -R.z), radiance_level).rgb;
+
+    // We always sample the first cubemap to achieve the required control flow uniformity (because of mip levels).
```
It would be better to use gradients and calculate the mip level. This is something that we need to do generally. I think maybe @JMS55 did it for the meshlet effort?

JMS55 replied:
I don't quite understand what the mip level issue is here. But for the diffuse map, there's only one mipmap level anyways, so feel free to use textureSampleLevel(mip: 0). For the specular map, you need to sample a specific mip based on roughness.

For the meshlet stuff, the gradients I did calculate manually, and patched pbr_fragment.wgsl to sample material textures using them. I don't _think_ there's any other textures I need to change to use explicit gradients samples as all the screen space textures are a single mip level anyways (except for the specular env map, which is different regardless). The only exception is shadow biasing, once we merge #10188.

robtfm replied:
i don't think anything forces diffuse maps to have no mips.

JMS55 replied:
The diffuse map should be a tiny 64x64 map at best. There's no reason to have mips - it's just a small LUT. If a user provides one with mips, we should ignore it.

pcwalton replied:
I use level 0.0 now and removed the comment.

superdump commented on crates/bevy_pbr/src/environment_map/mod.rs:
```diff
@@ -29,90 +59,460 @@ impl Plugin for EnvironmentMapPlugin {
+            return;
+        };
+
+        // Make sure the environment map is compatible with the ones we're
```
```suggestion
        // Make sure the environment map is compatible with the ones we've
```

superdump commented on crates/bevy_render/src/texture/image.rs:
```diff
@@ -186,7 +186,9 @@ impl Default for Image {
-                usage: wgpu::TextureUsages::TEXTURE_BINDING | wgpu::TextureUsages::COPY_DST,
+                usage: wgpu::TextureUsages::TEXTURE_BINDING
+                    | wgpu::TextureUsages::COPY_DST
+                    | wgpu::TextureUsages::COPY_SRC,
```
I'm not aware of any problems of adding COPY_SRC generally, but I do wonder if there is any.

JMS55 replied:
Afaik we shouldn't be adding usages we don't need, it'll hurt perf. I'd like to avoid this. Same with COPY_DST.

superdump reviewed (commented):
The code looks clean. I'm a bit too tired at this moment to thoroughly review it all.

I see we're adding a couple of large textures. Naively asking - are these necessary or can we reuse existing textures? Can we make them smaller?

pcwalton commented:
> The code looks clean. I'm a bit too tired at this moment to thoroughly review it all.
> 
> I see we're adding a couple of large textures. Naively asking - are these necessary or can we reuse existing textures? Can we make them smaller?

Thanks for the review! I'll see about reusing the existing skybox in the tree. That would eliminate one of the textures.

pcwalton commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -111,3 +111,15 @@ struct ClusterOffsetsAndCounts {
 };
 #endif
+
+struct ReflectionProbe {
```
"Light probe" is like the superclass that encompasses reflection probes and irradiance volumes (#10268), and possibly in the future reflection planes. I should make that more clear in the comments and commit :)

robtfm replied:
that makes sense if we're adding other probe types into the LightProbes struct in future (not into a separate uniform?), and will be clear enough then, don't think you need any extra comments.

pcwalton commented on crates/bevy_pbr/src/environment_map/mod.rs:
```diff
@@ -29,90 +59,460 @@ impl Plugin for EnvironmentMapPlugin {
+
+        let (width, height) = (dest_image.size.x as u32, dest_image.size.y as u32);
+
+        let mut command_encoder = render_device.create_command_encoder(&CommandEncoderDescriptor {
```
I like this idea and I can work on adding it. I think it makes sense to land it initially since it's an infrastructural change and adding `COPY_SRC` is ugly.

JMS55 replied:
Agree with rob. Will skip reviewing this part of the PR for now.

pcwalton replied:
As discussed on Discord this is now using binding arrays, so the extra GPU copies, as well as `COPY_SRC`, are gone.

JMS55 commented on crates/bevy_pbr/src/environment_map/environment_map.wgsl:
```diff
@@ -16,14 +18,46 @@ fn environment_map_light(
-    let radiance = textureSampleLevel(bindings::environment_map_specular, bindings::environment_map_sampler, vec3(R.xy, -R.z), radiance_level).rgb;
+
+    // We always sample the first cubemap to achieve the required control flow uniformity (because of mip levels).
+    let irradiance = textureSample(bindings::environment_map_diffuse, bindings::environment_map_sampler, vec3(N.xy, -N.z), max(cubemap_index, 0)).rgb;
```
Yeah I don't get why we're doing max(0) here, or what the comment about uniformity means...

pcwalton replied:
This is no longer applicable.

JMS55 commented on crates/bevy_pbr/src/environment_map/mod.rs:
```diff
@@ -26,67 +58,451 @@ impl Plugin for EnvironmentMapPlugin {
         app.register_type::<EnvironmentMapLight>()
             .add_plugins(ExtractComponentPlugin::<EnvironmentMapLight>::default());
+
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
```
This needs to be moved to finish().

JMS55 commented on crates/bevy_pbr/src/environment_map/mod.rs:
```diff
@@ -26,67 +58,451 @@ impl Plugin for EnvironmentMapPlugin {
+/// texture array at runtime. A warning is emitted if this restriction isn't followed.
 #[derive(Component, Reflect, Clone, ExtractComponent)]
-#[extract_component_filter(With<Camera3d>)]
+#[extract_component_filter(Or<(With<Camera3d>, With<LightProbe>)>)]
```
I think we can use `AnyOf` here.

pcwalton replied:
This uses `ExtractInstance` now so not applicable anymore

JMS55 commented on crates/bevy_pbr/src/environment_map/mod.rs:
```diff
@@ -26,67 +58,451 @@ impl Plugin for EnvironmentMapPlugin {
 pub struct EnvironmentMapLight {
+    /// The blurry, often-low-resolution cubemap representing diffuse light.
     pub diffuse_map: Handle<Image>,
+    /// The sharp, high-resolution cubemap representing specular light.
```
Eh that's misleading, the specular mips will also be blurry. Also applies to below.

JMS55 commented on examples/3d/reflection_probes.rs:
```diff
@@ -0,0 +1,270 @@
+//! This example shows how to place reflection probes in the scene.
```
Could use an extra toggle to pause the camera movement.

JMS55 commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -108,3 +108,15 @@ struct ClusterOffsetsAndCounts {
 #endif
+
+struct ReflectionProbe {
+    inverse_transform: mat4x4<f32>,
```
Can maybe be packed with affine to square or whatever to save some space? Idk

pcwalton replied:
I switched this to the transpose to save space.

JMS55 commented on crates/bevy_pbr/src/render/mesh_view_bindings.rs:
```diff
@@ -441,29 +455,32 @@ pub fn prepare_mesh_view_bind_groups(
-            let env_map_bindings =
-                environment_map::get_bindings(environment_map, &images, &fallback_cubemap);
+            // This needs to go up here to ensure correct drop order (otherwise
+            // the borrow check complains).
```
```suggestion
            // the borrow checker complains).
```

pcwalton replied:
Removed this stuff

JMS55 commented on crates/bevy_pbr/src/material.rs:
```diff
@@ -1,4 +1,9 @@
-use crate::*;
+use crate::{
```
Use the glob import please

JMS55 commented on crates/bevy_pbr/src/light_probe.rs:
```diff
@@ -0,0 +1,276 @@
+///
+/// Because the fragment shader does a linear search through the list for each
+/// fragment, this number needs to be relatively small.
+pub const MAX_VIEW_REFLECTION_PROBES: usize = 4;
```
Agree we could go up to something like 8/16.

pcwalton replied:
I bumped it to 8.

JMS55 commented on crates/bevy_pbr/src/light_probe.rs:
```diff
@@ -0,0 +1,276 @@
+/// fragment, this number needs to be relatively small.
+pub const MAX_VIEW_REFLECTION_PROBES: usize = 4;
+
+/// Adds support for light probes, cuboid bounding regions that apply global
```
```suggestion
/// Adds support for light probes: cuboid bounding regions that apply global
```

JMS55 commented on crates/bevy_pbr/src/light_probe.rs:
```diff
@@ -0,0 +1,276 @@
+    #[inline]
+    fn default() -> Self {
+        Self {
+            half_extents: Vec3A::splat(1.0),
```
Agreed

JMS55 commented on crates/bevy_pbr/src/light_probe.rs:
```diff
@@ -0,0 +1,276 @@
+    fn build(&self, app: &mut App) {
+        app.register_type::<LightProbe>();
+
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else {
```
Render app stuff needs to be moved to finish().

JMS55 commented on crates/bevy_pbr/src/light_probe.rs:
```diff
@@ -0,0 +1,276 @@
+                Render,
+                upload_light_probes
+                    .in_set(RenderSet::PrepareResources)
+                    .after(environment_map::prepare_environment_maps),
```
Agreed

pcwalton replied:
I moved both into `light_probe`.

JMS55 commented on crates/bevy_pbr/src/environment_map/mod.rs:
```diff
@@ -26,67 +58,451 @@ impl Plugin for EnvironmentMapPlugin {
+    }
+
+    // Ensures that the cubemap we're about to add to an array is compatible
+    // with the existing ones we've seen.  Returns true if it is and false if it
```
```suggestion
    // with the existing ones we've seen. Returns true if compatible, and false if not compatible [...]
```

JMS55 commented on examples/3d/reflection_probes.rs:
```diff
@@ -0,0 +1,270 @@
+// Creates the reflection probe, which is a box that refers to an environment
+// map and encloses the objects that are to be affected by it (in this case, the
+// sphere).
+fn create_reflection_probe(commands: &mut Commands, cubemaps: &Cubemaps) {
```
Please add a ReflectionProbeBundle, consisting of spatial bundle, light probe, and environment map light.

JMS55 reviewed (changes requested):
I'd like to not merge this PR until #9414 is first merged, and potentially also #8407 (for the EnvironmentMapLight exposure support).

pcwalton commented:
> I'd like to not merge this PR until #9414 is first merged, and potentially also #8407 (for the EnvironmentMapLight exposure support).

I reviewed #9414 but I don't think we should block this PR on that, since my PR doesn't care about how the reflection probe is generated.

pcwalton commented on Cargo.toml:
```diff
@@ -2356,6 +2356,17 @@ name = "fallback_image"
+name = "Reflection Probes"
+description = "Demonstrates reflection probes"
+category = "3D Rendering"
+wasm = true
```
`wasm = false` now

pcwalton commented on examples/3d/reflection_probes.rs:
```diff
@@ -0,0 +1,270 @@
+        },
+        LightProbe {
+            // 1.1 because the sphere's radius is 1.0 and we want to fully enclose it.
+            half_extents: Vec3A::splat(1.1),
```
I can't reproduce this anymore.

pcwalton commented on crates/bevy_pbr/src/environment_map/environment_map.wgsl:
```diff
@@ -16,14 +18,46 @@ fn environment_map_light(
+        let reflection_probe = light_probes.reflection_probes[reflection_probe_index];
+        let probe_space_pos =
+            (reflection_probe.inverse_transform * vec4<f32>(world_position, 1.0)).xyz;
+        if (all(abs(probe_space_pos) <= reflection_probe.half_extents)) {
```
Done.

pcwalton commented on crates/bevy_pbr/src/light_probe.rs:
```diff
@@ -0,0 +1,276 @@
+
+    /// The index of the environment map in the diffuse and specular cubemap
+    /// texture arrays.
+    cubemap_index: i32,
```
Not if there is no view cubemap index. In that case it will be -1.

pcwalton commented:
This is ready for re-review.

pcwalton commented:
I now only add one texture. This is the minimum, as a reflection probe has to reflect some geometry in the scene to be interesting, and there's no such texture in the tree already.

JMS55 commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -109,3 +109,18 @@ struct ClusterOffsetsAndCounts {
+    // This must match `MAX_VIEW_REFLECTION_PROBES` on the Rust side.
+    reflection_probes: array<ReflectionProbe, 8u>,
+    reflection_probe_count: i32,
+    // -1 if no cubemap is associated.
```
What is the separate cubemap index for? The view environment map right? Can you clarify the name?

pcwalton replied:
I renamed it to `view_cubemap_index`.

JMS55 commented on crates/bevy_pbr/src/render/mesh.rs:
```diff
@@ -837,6 +844,9 @@ impl SpecializedMeshPipeline for MeshPipeline {
             },
         ));
 
+        #[cfg(all(not(feature = "shader_format_glsl"), not(target_arch = "wasm32")))]
```
Why does GLSL matter? What is the issue with enabling it?

pcwalton replied:
GLSL doesn't support bindless and it was crashing in Naga. I discussed with @robtfm on Discord and we came to the conclusion that we can just disable on GLSL.

JMS55 commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -968,10 +962,7 @@ pub fn prepare_lights(
-                .and_then(|env_map| images.get(&env_map.specular_map))
-                .map(|specular_map| specular_map.mip_level_count - 1)
-                .unwrap_or(0),
+            environment_map_smallest_specular_mip_level: 0, /* FIXME */
```
This needs to be fixed. Specular envmaps won't look correct at all. What prevented you from doing so?

pcwalton replied:
So I ended up just using `textureNumLevels()` as the comment in `environment_map.wgsl` suggests. Since reflection probes are disabled on WebGL 2, we don't have to worry about compatibility for those. I did need to add a `smallest_specular_mip_level_for_view` field to the `LightProbes` struct though, since that one *is* used on WebGL 2.

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.wgsl:
```diff
@@ -0,0 +1,121 @@
+            reflection_probe_index += 1) {
+        let reflection_probe = light_probes.reflection_probes[reflection_probe_index];
+
+        // Transpose the inverse transpose transform to recover the inverse
```
Nit: Change just comment to just "Unpack the inverse transform"

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.wgsl:
```diff
@@ -0,0 +1,121 @@
+    // because textureNumLevels() does not work on WebGL2
+    let radiance_level = perceptual_roughness * f32(bindings::lights.environment_map_smallest_specular_mip_level);
+
+#ifdef ENVIRONMENT_MAP_LIGHT_PROBES
```
I think we should rename this shader def to `MULTIPLE_LIGHT_PROBES_IN_ARRAY` or something like that.

pcwalton replied:
Done

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.rs:
```diff
@@ -0,0 +1,391 @@
+//!
+//! Currently, reflection probes (i.e. environment maps attached to light
+//! probes) use binding arrays (also known as bindless textures) and
+//! consequently aren't supported on WebGL 2 or WebGPU. Reflection probes are
```
```suggestion
//! consequently aren't supported on WebGL2 or WebGPU. Reflection probes are
```

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.rs:
```diff
@@ -0,0 +1,391 @@
+//! Environment maps and reflection probes.
```
The old docs contained some info about the split sum approximation, and talked about the use case for view-level envmaps (better quality dynamic indirect lighting compared to AmbientLight). I'd like to keep that info.

pcwalton replied:
OK, I merged the text in. Let me know if it looks good.

JMS55 replied:
Lgtm, thanks

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.rs:
```diff
@@ -0,0 +1,391 @@
+    pub(crate) specular: AssetId<Image>,
+}
+
+/// A convenient bundle that contains everything needed to make an entity a
```
```suggestion
/// A bundle that contains everything needed to make an entity a
```

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.rs:
```diff
@@ -0,0 +1,391 @@
+    ///
+    /// This is a vector of `wgpu::TextureView`s. But we don't want to import
+    /// `wgpu` in this crate, so we refer to it indirectly like this.
+    diffuse_texture_views: Vec<&'a <TextureView as std::ops::Deref>::Target>,
```
Nit: I'd import Deref here instead of fully qualifying it.

JMS55 commented on crates/bevy_pbr/src/light_probe/environment_map.rs:
```diff
@@ -0,0 +1,391 @@
+#[cfg(all(not(feature = "shader_format_glsl"), not(target_arch = "wasm32")))]
+impl RenderViewEnvironmentMaps {
+    /// Returns true if there are no environment maps for this view or false if
+    /// there are such environment maps.
```
Doc comment wording is a bit confusing. You can just say whether there are any envmaps associated with the view, or something.

JMS55 commented on crates/bevy_pbr/src/light_probe/mod.rs:
```diff
@@ -0,0 +1,389 @@
+/// cubemaps applied to all objects that a view renders.
+pub struct LightProbePlugin;
+
+/// A cuboid region that provides global illumination to all fragments inside it.
```
Nit: Start the doc comment with something like "a marker component for a light probe, which is a ..."

JMS55 commented on crates/bevy_pbr/src/light_probe/mod.rs:
```diff
@@ -0,0 +1,389 @@
+            Shader::from_wgsl
+        );
+
+        app.register_type::<LightProbe>();
```
We should register EnvironmentMapLight as well.

JMS55 reviewed (changes requested):
Overall looks good, but needs some testing to ensure nothing is broken on WebGL2/WebGPU/etc. Easy potential for breaking things by accident.

The GLSL changes seem strange to me. What's the issue? It would also be a bit nicer if you could pull the parts that can out to a separate PR.

JMS55 commented:
Not for this PR, but so that it doesn't get lost, lots of good baked GI techniques here: https://advances.realtimerendering.com/s2015/SIGGRAPH_2015_Remedy_Notes.pdf

pcwalton commented:
I addressed the comments. The issue with WebGL 2 is simply that neither WebGL 2 nor Naga's implementation of GLSL supports bindless textures.

I'm not sure which could be pulled out to a separate PR--could you elaborate?

JMS55 commented:
> I addressed the comments. The issue with WebGL 2 is simply that neither WebGL 2 nor Naga's implementation of GLSL supports bindless textures.

Right, but we don' use any GLSL shaders, so I'm wondering what the issue is. Do user GLSL shaders somehow cause issues?

pcwalton commented:
> > I addressed the comments. The issue with WebGL 2 is simply that neither WebGL 2 nor Naga's implementation of GLSL supports bindless textures.
> 
> Right, but we don't use any GLSL shaders, so I'm wondering what the issue is. Do user GLSL shaders somehow cause issues?

Ah, yeah, that's a weird `naga_oil` quirk.

@superdump asked this in Discord:

> I was going to ask why your situation means glsl is being generated. @robtfm we don’t generate glsl up-front in oil for imports do we?


@robtfm wrote in Discord:

> for modules we do, not for top-level shaders
> we rebuild importable modules in each supported language to make sure all the used idents are valid in that language
> at least, in debug. @pcwalton does it work in release?

Indeed it does work in release. So the issue seems to be that this debugging feature for GLSL causes problems in debug even when only WGSL is used.

JMS55 commented:
Is it intended that the view cubemap and the light probe maps must all have the same amount of specular mip levels? You're using the lowest specular mip parameter from the view for the light probes as well.

JMS55 commented:
For both the view cubemap and any light probes, if the mesh has a lightmap, we should probably ignore the diffuse lighting from cubemaps. Otherwise, we'd double count it.

pcwalton commented:
> Is it intended that the view cubemap and the light probe maps must all have the same amount of specular mip levels? You're using the lowest specular mip parameter from the view for the light probes as well.

Could you show me where in the code I do this? The "smallest specular mip for the view" value is used at https://github.com/bevyengine/bevy/pull/10057/files#diff-c13ab0c48045d7a1b51300adb87e4f7fd247e715b5991d54ee2e6f27e2b45f97R87 which is only enabled if `MULTIPLE_LIGHT_PROBES_IN_ARRAY` is *not* present, as it's behind an `#else`. For light probe maps, we ignore that value and use `textureNumLevels` to check the number of mips directly: https://github.com/bevyengine/bevy/pull/10057/files#diff-c13ab0c48045d7a1b51300adb87e4f7fd247e715b5991d54ee2e6f27e2b45f97R64

pcwalton commented:
> For both the view cubemap and any light probes, if the mesh has a lightmap, we should probably ignore the diffuse lighting from cubemaps. Otherwise, we'd double count it.

Yep, now that lightmaps have landed I made that change.

JMS55 commented:
Nvm, I missed that you were using textureNumLevels for the light probes.

mockersf commented:
it also broke iOS:
```
wgpu error: Validation Error

Caused by:
    In Device::create_bind_group_layout
      note: label = `mesh_view_layout`
    Binding 14 entry is invalid
    Features Features(TEXTURE_BINDING_ARRAY) are required but not enabled on the device
```

mockersf commented:
and Android:
```
wgpu error: Validation Error
Caused by:
    In Device::create_bind_group_layout
      note: label = `mesh_view_layout`
    Too many bindings of type SampledTextures in Stage ShaderStages(FRAGMENT), limit is 16, count was 21
```
